### PR TITLE
deps: enabled building of node on zOS.

### PR DESCRIPTION
**Pull request by exinfinitum**
The changes contained within this pull request allows us to build Node.js on a zOS platform. In order to do this, perform the following actions:

    In deps/openssl/openssl, run "./Configure OS390-Unix64", then run "make depend".
    In the root node directory, run "./configure --without-snapshot", followed by "make -j7".

At present, only the 64 bit version has been tested to be working. 32 bit support is not guaranteed.

A number of hacks were needed to get our code to work properly. These hacks are temporary solutions for certain issues (mostly compiler-related) and are documented below.

configure:449-450: The node build configure system has a sophisticated method of figuring out which platform we are building on. This works by getting the OS platform name and then figuring out which compiler to use; and then using the compiler to get the host architecure. This system currently is not configured to accept either xlc or zos, so we use a temporary workaround to simply return the s390x architecture should zOS be the OS.

src/node.cc:762, 764, 824, 826, 894, 896, 2081: Exception::Error results in compile errors under xlc due to compiler-side issues (compiler couldn't find Exception, despite it being in the v8 namespace, which we are using). To solve this, it was necessary to explicitly define the namespace of Exception as being in v8, via v8::Exception::Error.

src/node_crypto.cc:839, 1680, 1933, 1978, 2001, 4486, 4747: See above on node.cc.

src/node_http_parser.cc:427: See above on node.cc.

src/tls_wrap.cc:431, 808: See above on node.cc.

src/node.h:381-390, 394-396, 432: Using a static variable function was necessary since xlc does not support GCC's "constructor" and "destructor" functions, which are evaluated before/after the main function respectively. Don't know if this qualifies as a hack per se, since the static variable method is one that is supported by the C++ standard, whereas GCC constructor/destructor are not.

v8z-patch:8-17: Using private functions, the v8 SealHandleScope makes it difficult to create heap-allocated or illegal handle scopes through including a number of private functions which were declared but not defined (no code is provided). Unfortunately, this practice is not supported by xlc, and therefore this block of code had to be removed.

v8z-patch itself could be considered a hack as well. Once node is ready for release and a finalized v8zos is packaged with node, this file can be removed.

### CODE REVIEW

File: deps/http_parser/http_parser.gyp [hunk 1]
```diff
@@ -12,7 +12,7 @@
       #       RuntimeLibrary MUST MATCH across the entire project
       'Debug': {
         'defines': [ 'DEBUG', '_DEBUG' ],
-        'cflags': [ '-Wall', '-Wextra', '-O0', '-g', '-ftrapv' ],
+        'cflags': [ '-O0', '-g', '-ftrapv' ],
```
Comment (richardlau) [inline deps/http_parser/http_parser.gyp]: Maybe keep the defaults and negate for `os390`?
e.g. 
```
      ['OS == "os390"', {
        'cflags!': [ '-Wall', '-Wextra', ],
      }],
```

File: deps/http_parser/http_parser.gyp [hunk 2]
```diff
@@ -21,7 +21,7 @@
       },
       'Release': {
         'defines': [ 'NDEBUG' ],
-        'cflags': [ '-Wall', '-Wextra', '-O3' ],
+        'cflags': [ '-O3' ],
```
Comment (richardlau) [inline deps/http_parser/http_parser.gyp]: Maybe keep the defaults and negate for `os390`?
e.g. 
```
      ['OS == "os390"', {
        'cflags!': [ '-Wall', '-Wextra', ],
      }],
```

File: deps/openssl/openssl.gyp [hunk 3]
```diff
@@ -1125,7 +1125,6 @@
           'ENGINESDIR="/dev/null"',
           'TERMIOS',
         ],
-        'cflags': ['-Wno-missing-field-initializers'],
```
Comment (richardlau) [inline deps/openssl/openssl.gyp]: Again, keep and negate for `os390`
e.g. 
```
      ['OS == "os390"', {
        'cflags!': [ '-Wno-missing-field-initializers' ],
      }],
```

File: deps/uv/uv.gyp [hunk 4]
```diff
@@ -10,15 +10,19 @@
...
     ],
     'xcode_settings': {
-      'WARNING_CFLAGS': [ '-Wall', '-Wextra', '-Wno-unused-parameter' ],
-      'OTHER_CFLAGS': [ '-g', '--std=gnu89', '-pedantic' ],
+      'conditions':[
+        ['OS!="os390"', {
+          'WARNING_CFLAGS': [ '-Wall', '-Wextra', '-Wno-unused-parameter' ],
+          'OTHER_CFLAGS': [ '-g', '--std=gnu89', '-pedantic' ],
+        }],
+      ]
```
Comment (richardlau) [inline deps/uv/uv.gyp]: Is this change necessary? I would expect `xcode_settings` to only apply to OS X.

File: tools/gyp/pylib/gyp/generator/make.py [hunk 5]
```diff
@@ -310,7 +310,7 @@ def CalculateGeneratorInputInfo(params):
 # We write to a dep file on the side first and then rename at the end
 # so we can't end up with a broken dep file.
 depfile = $(depsdir)/$(call replace_spaces,$@).d
-DEPFLAGS = -qmakedep -MF $(depfile).raw
+DEPFLAGS = -qmakedep -qLONGLONG -qLANGLVL=EXTENDED -D_UNIX03_THREADS -D_UNIX03_SOURCE -D_OPEN_SYS_IF_EXT=1 -qCHARS=signed -qXPLINK -qFLOAT=IEEE -D_OPEN_MSGQ_EXT -D_XOPEN_SOURCE_EXTENDED -D_ALL_SOURCE -D_LARGE_TIME_API -DUTMP_FILE=\"/etc/utmpx\" -DPATH_MAX=255 -D_OPEN_SYS_SOCK_IPV6 -D_OPEN_SYS_FILE_EXT -DIP_TTL=-1 -DNI_MAXHOST=1024 -DNI_MAXSERV=32 -MF $(depfile).raw
```
Comment (richardlau) [inline tools/gyp/pylib/gyp/generator/make.py]: Will this affect building on non-zOS ?

Comment (exinfinitum) [inline tools/gyp/pylib/gyp/generator/make.py]: This is not from the latest commit. The latest commit (titled "Linux compatibility") will use the appropriate flags for zOS and non-zOS platforms.

**Comment by exinfinitum** [comment]
Updated.

File: deps/openssl/openssl.gyp [hunk 6]
```diff
@@ -1146,8 +1147,8 @@
       ['is_clang==1 or gcc_version>=43', {
         'cflags': ['-Wno-old-style-declaration'],
       }],
-      ['OS not in "win os390"', {
-        'cflags': ['-Wno-missing-field-initializers'],
+      ['OS in "win os390"', {
+        'cflags!': ['-Wno-missing-field-initializers'],
```
Comment (richardlau) [inline deps/openssl/openssl.gyp]: Don't need win here if using negation.

**Comment by exinfinitum** [comment]
Updated. This commit lets us go to prompt on zOS, but there are still many blocking issues before this prompt becomes practical (e.g. console does not output errors, 100% test failure rate).

File: configure [hunk 7]
```diff
@@ -494,7 +496,7 @@ def compiler_version():
   try:
     proc = subprocess.Popen(shlex.split(CC) + ['--version'],
                             stdout=subprocess.PIPE)
-  except WindowsError:
+  except OSError:
```
Comment (thinkingdust) [inline configure]: Should this still handle WindowsError, i.e.
except (WindowsError, OSError):

File: deps/uv/vcbuild.bat [hunk 8]
```diff
@@ -1,153 +1,153 @@
...
-:help
-echo vcbuild.bat [debug/release] [test/bench] [clean] [noprojgen] [nobuild] [x86/x64] [static/shared]
-echo Examples:
-echo   vcbuild.bat              : builds debug build
-echo   vcbuild.bat test         : builds debug build and runs tests
-echo   vcbuild.bat release bench: builds release build and runs benchmarks
-goto exit
-
-:exit
+@echo off
```
Comment (thinkingdust) [inline deps/uv/vcbuild.bat]: Looks like a line ending issue in this .bat file, we probably don't want to commit this change.

File: tools/gyp/pylib/gyp/generator/make.py [hunk 9]
```diff
@@ -142,7 +142,7 @@ def CalculateGeneratorInputInfo(params):
 # special "figure out circular dependencies" flags around the entire
 # input list during linking.
 quiet_cmd_link = LINK($(TOOLSET)) $@
-cmd_link = $(LINK.$(TOOLSET)) $(GYP_LDFLAGS) $(LDFLAGS.$(TOOLSET)) -o $@ -Wl,--start-group $(LD_INPUTS) -Wl,--end-group $(LIBS)
+cmd_link = $(LINK.$(TOOLSET)) $(GYP_LDFLAGS) $(LDFLAGS.$(TOOLSET)) -o $@ $(LD_INPUTS) $(LIBS)
```
Comment (thinkingdust) [inline tools/gyp/pylib/gyp/generator/make.py]: Should this be if sys.platform.startswith('os390')?

File: tools/js2c.py [hunk 10]
```diff
@@ -53,8 +53,8 @@ def ToCArray(filename, lines):
 
     value = ord(chr)
 
-    if value >= 128:
-      print 'non-ascii value ' + filename + ':' + str(row) + ':' + str(col)
+    if value >= 256:
```
Comment (thinkingdust) [inline tools/js2c.py]: if sys.platform.startswith('os390')?

File: configure (hunk 7 above)
Comment (exinfinitum) [inline configure]: According to http://stackoverflow.com/questions/8890010/os-rename-on-windows-raise-windowserror-instead-of-oserror, WindowsError is a subclass of OSError, so it will be caught by the except.

File: tools/gyp/pylib/gyp/generator/make.py (hunk 9 above)
Comment (exinfinitum) [inline tools/gyp/pylib/gyp/generator/make.py]: Whoops, forgot to properly revert this when trying to allow linux compatibility. Fixed in next update.

File: tools/js2c.py (hunk 10 above)
Comment (exinfinitum) [inline tools/js2c.py]: Fixed in latest commit

**Comment by exinfinitum** [comment]
Note: v8 replacements have been split off into another PR (#64).

File: deps/v8z/include/v8.h [hunk 11]
```diff
@@ -943,6 +943,24 @@ class ScriptOrigin {
...
+  SealHandleScope(const SealHandleScope&);
+  void operator=(const SealHandleScope&);
+  void* operator new(size_t size);
+  void operator delete(void*, size_t);
+
+  internal::Isolate* isolate_;
+  int prev_level_;
+  internal::Object** prev_limit_;
+};
+
```
Comment (thinkingdust) [inline deps/v8z/include/v8.h]: Isn't this done by the v8z-patch so it does not need to be committed?

File: deps/v8z/src/api.cc [hunk 12]
```diff
@@ -669,6 +669,15 @@ void Context::Enter() {
...
 
+void Isolate::SetAbortOnUncaughtException(
+  abort_on_uncaught_exception_t callback) {
+  i::Isolate* isolate = reinterpret_cast<i::Isolate*>(this);
+  isolate->SetAbortOnUncaughtException(callback);
+}
+
+
+
+
```
Comment (thinkingdust) [inline deps/v8z/src/api.cc]: Isn't this done by the v8z-patch so it does not need to be committed?

**Comment by exinfinitum** [comment]
Updated.

**Comment by exinfinitum** [comment]
Squashed.

File: deps/v8z/build/toolchain.gypi [hunk 13]
```diff
@@ -520,6 +520,24 @@
...
+          '-qbitfield=signed',
+          '-qnortti',
+          '-Wc,expo',
+          '-g',
+          '-D__IBMCPP_TR1__',
+          '-D__BIG_ENDIAN=4321',
+          '-D__BYTE_ORDER=__BIG_ENDIAN',
+          '-D_OPEN_SYS_TIMED_EXT=1',
+        ],
+      }],
```
Comment (thinkingdust) [inline deps/v8z/build/toolchain.gypi]: Won't this change appear in your other commit which updates the v8z directory?

File: deps/zlib/zlib.gyp [hunk 14]
```diff
@@ -74,6 +74,12 @@
                 'USE_FILE32API'
               ],
             }],
+            ['OS=="os390"', {
+              # Mac, Android and the BSDs don't have fopen64, ftello64, or
+              # fseeko64. We use fopen, ftell, and fseek instead on these
+              # systems.
+              'cflags': [ '-D_XOPEN_SOURCE_EXTENDED=1', '-D_XOPEN_SOURCE=600' ], 
+            }]
```
Comment (thinkingdust) [inline deps/zlib/zlib.gyp]: Comment seems unrelated.

File: src/node.cc [hunk 15]
```diff
@@ -759,9 +764,9 @@ Local<Value> ErrnoException(Isolate* isolate,
         String::Concat(cons3, String::NewFromUtf8(env->isolate(), path));
     Local<String> cons5 =
         String::Concat(cons4, FIXED_ONE_BYTE_STRING(env->isolate(), "'"));
-    e = Exception::Error(cons5);
+    e = v8::Exception::Error(cons5);
```
Comment (mhdawson) [inline src/node.cc]: Why is adding the v8 namespace required here ?

File: v8z-patch [hunk 16]
```diff
@@ -0,0 +1,129 @@
...
+@@ -1341,6 +1344,8 @@
+ 
+   v8::Isolate::UseCounterCallback use_counter_callback_;
+ 
++  abort_on_uncaught_exception_t abort_on_uncaught_exception_callback_;
++
+   friend class ExecutionAccess;
+   friend class HandleScopeImplementer;
+   friend class IsolateInitializer;
+
```
Comment (mhdawson) [inline v8z-patch]: I'm guessing this file should not be in the patch at all.

File: src/node.cc (hunk 15 above)
Comment (joransiu) [inline src/node.cc]: That's to temporarily workaround a known issue with xlC.

Comment (exinfinitum) [inline src/node.cc]: XLC apparently has trouble finding "Exception::Error" even though we're "using namespace v8", resulting in compile errors. It was necessary to explicitly use the v8 namespace to resolve the errors.

File: v8z-patch (hunk 16 above)
Comment (exinfinitum) [inline v8z-patch]: V8 from the ibmruntimes/v8z repo will not work as a drop-in replacement for the v8 dependency packaged in the node repository. This patch contains the necessary changes in order to allow a fresh checkout of v8z from ibmruntimes/v8z:3.28-zos to function properly when used to replace our existing v8z folder. It has been included due to the fact that v8zos is being updated constantly right now.

This file should be removed when node is ready for release, as by then v8zos would probably be finalized as well.

File: deps/zlib/zlib.gyp (hunk 14 above)
Comment (exinfinitum) [inline deps/zlib/zlib.gyp]: Comment removed.

**Comment by exinfinitum** [comment]
@mhdawson @thinkingdust Updated.

File: src/node.cc (hunk 15 above)
Comment (mhdawson) [inline src/node.cc]: I'm just thinking from the point of getting the changes into the community. This is one that might be a harder sell.  But I guess there is a change will have a fix on the compiler side by the time we need it to go in.

Comment (joransiu) [inline src/node.cc]: I believe @jBarz already has opened an issue with xlC team for this problem.

File: src/node.h [hunk 17]
```diff
@@ -418,9 +429,7 @@ extern "C" NODE_EXTERN void node_module_register(void* mod);
       priv,                                                           \
       NULL                                                            \
     };                                                                \
-    NODE_C_CTOR(_register_ ## modname) {                              \
-      node_module_register(&_module);                                 \
-    }                                                                 \
+    NODE_C_CTOR(_register_ ## modname)                                \                                  \
```
Comment (thinkingdust) [inline src/node.h]: This is causing a problem building on Linux.
../src/async-wrap.cc:190:1: error: stray ▒\▒ in program
NODE_MODULE_CONTEXT_AWARE_BUILTIN(async_wrap, node::Initialize)
^
Maybe add an if MVS ... else ... here.

**Comment by exinfinitum** [comment]
@thinkingdust Updated.

**Comment by jBarz** [comment]
LGTM
